File: Year-2019/Apr/2019-04-01-160/watercow/160.c
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode * getIntersectionNode(ListNode *headA, ListNode *headB) {
		if (headA == NULL || headB == NULL) return NULL;

		// 首先让headB链表构成一个环
		ListNode * p0 = headB;
		while (p0->next != NULL) {
			p0 = p0->next;
		}
		p0->next = headB;


		// 剩下的工作就是找入口环的入口点
		ListNode * p1 = headA;
		ListNode * p2 = headA;
		while (p1 != NULL && p1->next != NULL) {
			p2 = p2->next;
			p1 = p1->next->next;
			if (p1 == p2) {
				p2 = headA;
				while (p2 != p1) {
					p2 = p2->next;
					p1 = p1->next;
				}
				p0->next = NULL;
				return p1;
			}
		}
		p0->next = NULL;
		return NULL;
	}
};

File: Year-2019/Apr/2019-04-01-160/lb/160.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int ALength = 0;
        int BLength = 0;
        int AStarte = 0;
        int BStarte = 0;
        ListNode node;
        node = headA;
        while (node != null){
            ALength ++;
            node = node.next;
        }
        node = headB;
        while (node != null){
            BLength++;
            node = node.next;
        }
        //使A和B链表统一长度，同一起跑点
        ListNode nodeA = headA;
        ListNode nodeB = headB;
        if (ALength > BLength){
            AStarte = ALength - BLength +1;
            for (int i =1;i<AStarte;i++){
                nodeA = nodeA.next;
            }
        }
        else{
            BStarte  = BLength - ALength+1;
            for (int i =1 ;i<BStarte;i++){
                nodeB = nodeB.next;
            }
        }
        //找交点
        node = null;
        while (nodeA != null){
            if (nodeA == nodeB)
                return nodeA;
             nodeA = nodeA.next;
             nodeB = nodeB.next;
        }
        return null;
    }
}
